atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/TracingKeyValueService.java

```diff
@@ -0,0 +1,349 @@
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.keyvalue.impl;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ForwardingObject;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Multimap;
+import com.palantir.atlasdb.keyvalue.api.BatchColumnRangeSelection;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.CheckAndSetException;
+import com.palantir.atlasdb.keyvalue.api.CheckAndSetRequest;
+import com.palantir.atlasdb.keyvalue.api.ColumnRangeSelection;
+import com.palantir.atlasdb.keyvalue.api.ColumnSelection;
+import com.palantir.atlasdb.keyvalue.api.KeyAlreadyExistsException;
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.api.RangeRequest;
+import com.palantir.atlasdb.keyvalue.api.RowColumnRangeIterator;
+import com.palantir.atlasdb.keyvalue.api.RowResult;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.api.Value;
+import com.palantir.atlasdb.tracing.CloseableTrace;
+import com.palantir.common.base.ClosableIterator;
+import com.palantir.util.paging.TokenBackedBasicResultsPage;
+
+/**
+ * Wraps a {@link KeyValueService}'s methods with {@link com.palantir.remoting1.tracing.Tracer}
+ * instrumentation.
+ */
+public final class TracingKeyValueService extends ForwardingObject implements KeyValueService {
+
+    private static final String SERVICE_NAME = "atlasdb-kvs";
+
+    private final KeyValueService delegate;
+
+    private TracingKeyValueService(KeyValueService delegate) {
+        this.delegate = Preconditions.checkNotNull(delegate, "delegate");
+    }
+
+    public static KeyValueService create(KeyValueService keyValueService) {
+        return new TracingKeyValueService(keyValueService);
+    }
+
+    private static CloseableTrace startLocalTrace(CharSequence operationFormat, Object... formatArguments) {
+        return CloseableTrace.startLocalTrace(SERVICE_NAME, operationFormat, formatArguments);
+    }
+
+    @Override
+    protected KeyValueService delegate() {
+        return delegate;
+    }
+
+    @Override
+    public void addGarbageCollectionSentinelValues(TableReference tableRef, Set<Cell> cells) {
+        //noinspection unused - try-with-resources closes trace
```

Comment:
I'm inclined to leave the comments at use site to avoid someone mistakenly "fixing the glitch" and removing the try-with-resources